src/source/src/table.rs

```diff
@@ -67,75 +69,134 @@ impl TableDmlHandle {
-            }
+    pub fn write_handle(&self, txn_id: TxnId) -> Result<WriteHandle> {
+        // The `changes_txs` should not be empty normally, since we ensured that the channels
+        // between the `TableDmlHandle` and the `SourceExecutor`s are ready before we making the
+        // table catalog visible to the users. However, when we're recovering, it's possible
+        // that the streaming executors are not ready when the frontend is able to schedule DML
+        // tasks to the compute nodes, so this'll be temporarily unavailable, so we throw an
+        // error instead of asserting here.
+        // TODO: may reject DML when streaming executors are not recovered.
+        let guard = self.core.read();
+        if guard.changes_txs.is_empty() {
+            return Err(RwError::from(anyhow!(
+                "no available table reader in streaming source executors"
+            )));
         }
+        let len = guard.changes_txs.len();
+        let tx = guard
+            .changes_txs
+            .get((txn_id % len as u64) as usize)
+            .context("no available table reader in streaming source executors")?
+            .tx
+            .clone();
+
+        Ok(WriteHandle::new(txn_id, tx))
     }
 
     /// Get the reference of all columns in this table.
     pub(super) fn column_descs(&self) -> &[ColumnDesc] {
         self.column_descs.as_ref()
     }
+
+    pub fn check_chunk_schema(&self, chunk: &StreamChunk) {
+        risingwave_common::util::schema_check::schema_check(
+            self.column_descs
+                .iter()
+                .filter_map(|c| (!c.is_generated()).then_some(&c.data_type)),
+            chunk.columns(),
+        )
+        .expect("table source write txn_msg schema check failed");
+    }
 }
 
-#[easy_ext::ext(TableDmlHandleTestExt)]
-impl TableDmlHandle {
-    /// Write a chunk and assert that the chunk channel is not blocking.
-    fn write_chunk_ready(&self, chunk: StreamChunk) -> Result<oneshot::Receiver<usize>> {
-        self.write_chunk(chunk).now_or_never().unwrap()
+#[derive(Debug, PartialEq)]
+enum TxnState {
+    Init,
+    Begin,
+    Committed,
+    Rollback,
+}
+
+/// [`WriteHandle`] writes its data into a table in a transactional way.
+///
+/// First, it needs to call `begin()` and then write chunks by calling `write_chunk()`.
+///
+/// Finally call `end()` to commit the transaction or `rollback()` to rollback the transaction.
+///
+/// If the [`WriteHandle`] is dropped with a `Begin` transaction state, it will automatically
+/// rollback the transaction.
+pub struct WriteHandle {
+    txn_id: TxnId,
+    tx: mpsc::UnboundedSender<(TxnMsg, oneshot::Sender<usize>)>,
+    // Indicate whether `TxnMsg::End` or `TxnMsg::Rollback` have been sent to the write channel.
+    txn_state: TxnState,
+}
+
+impl Drop for WriteHandle {
+    fn drop(&mut self) {
+        if self.txn_state == TxnState::Begin {
+            let _ = self.rollback();
+        }
+    }
+}
+
+impl WriteHandle {
```

Comment:
BTW, I've come up with another approach to do synchronization between batch and streaming DML executors:

1. BatchDml sends a `(channel_rx, oneshot_tx)` to StreamDml that indicates a transaction starts.
2. BatchDml continues sending chunks to `channel_tx`.
3. StreamDml receives chunks from the `channel_rx` of all concurrent writers with `select`.
4. BatchDml sends a control message `End` that indicates the transaction gracefully ends, then StreamDml notifies the transaction commit back with `oneshot_tx`.
5. If the batch jobs are cancelled or failed, StreamDml will get the channel closed unexpectly and clean up the buffers.

The underlying ideas are the same but only the differences on implementations. Never mind. 🤣

Reply:
I have a doubt. When `StreamDml` updates its concurrent `channel_rx`? If it is possible, we can remove the `Rollback` message.

Reply:
Should be possible by selecting the "channel stream" with a [`FuturesUnordered`](https://docs.rs/futures/latest/futures/prelude/stream/struct.FuturesUnordered.html#method.push), but seems to be a little bit tricky. The current implementation looks much more intuitive.